Feature/multiple transports doc (#15)
smartdevicelink/sdl_core_documentation: docs/Multiple Transports Implementation/index.md

# General Description

> The multiple transports feature allows apps connected to core to start another connection over a different transport for certain services(For example, the app connecting over bluetooth and then connecting over WiFi for video streaming). This guide will walk you through how to set up multiple transports using the `smartDeviceLink.ini` file.

Reviewer: I would include a space before parentheses.

Reviewer: ```suggestion
The multiple transports feature allows apps connected to Core to start another connection over a different transport for certain services (for example, an app connected over bluetooth using WiFi as a secondary transport for video streaming). This guide will walk you through how to configure the multiple transports feature using the `smartDeviceLink.ini` file.
```

## Implementation
- After Proxy is connected to Core, it initiates another connection over a different transport.
- Core tells Proxy which transport can be used as Secondary Transport.
- The services that are allowed on the Secondary Transport are specified by Core.

!!! Note

RPC and Hybrid services only run on the Primary Transport

!!!

> Three protocol control frames are used to implement multiple transports

Reviewer: ```suggestion
There are three protocol control frames which are used in the implementation of multiple transports
```

### `StartService ACK`
```
{
    	audioServiceTransports : [1, 2]
    	videoServiceTransports : [1, 2]
    	secondaryTransports : [TCP_WIFI]
}
```

Core responds to the proxy's `StartService` request with additional parameters audioServiceTransports, videoServiceTransports and secondaryTransports.   

- The secondaryTransports contains an array of the allowed secondary transports for the current primary transport. 

> - AudioServiceTransports and videoServiceTransports describe which service is allowed to run on which transports (Primary, Secondary or both). The Proxy uses this information and starts services only on an allowed transport.

Reviewer: ```suggestion
- `audioServiceTransports` and `videoServiceTransports` describe which services are allowed to run on which transports (Primary, Secondary, or both). The Proxy uses this information and starts services only on allowed transports.
```

- Since RPC and Hybrid services always run on Primary Transport, only Video and Audio services are configurable.

### `TransportEventUpdate`
```
{
    tcpIpAddress : 192.168.1.1
    tcpPort : 12345
}
```

> Core sends a TransportEventUpdate to the proxy to provide additional information required to connect to the secondary transport  

Reviewer: ```suggestion
Core sends a TransportEventUpdate to the proxy to provide additional information required to connect over the TCP transport
```

> - In the case of TCP as the secondary transport, the TransportEventUpdate contains the `tcpIpAddress` and `tcpPort` params.

Reviewer: Considering this is only used for TCP right now, I would just remove this bullet

- If the `tcpIpAddress` field is empty, the secondary transport is unavailable and proxy will not send a RegisterSecondaryTransport request 

### `RegisterSecondaryTransport`

Using the information in the StartService ACK and TransportEventUpdate frames, the proxy sends a RegisterSecondary transport request over the secondary transport.  

> - If core sends back a RegisterSecondaryTransport ACK, the proxy starts the service over the secondary transport

Reviewer: ```suggestion
- If Core sends back a RegisterSecondaryTransport ACK, the proxy can start services over the secondary transport
```

## Operation Examples
### Start Service(WiFi as secondary transport)  

|||

> Figure 1

Reviewer: I would just use the titles here instead of in a header for each diagram. So, instead of Figure 1, just put `Start Service (WiFi as secondary transport)` here and remove the header

![StartService](./assets/StartService.png)
|||

### Start Video/Audio service(Over secondary transport)  

|||
Figure 2
![StartServiceVideo](./assets/StartServiceVideo.png)
|||

### Start Video/Audio service(No transport available)  

|||
Figure 3

> ![StartServiceVideo NAK](./assets/StartServiceNAK_Video.png)

Reviewer: Something should be included to show which connection each message is sent over, since the app has two connections to Core.

Author: I specified the transport type in the description on the left and I also made changes to the arrows in the diagrams to differentiate between primary and secondary transports

|||

> ### Backwards Compatibility(New Proxy/Old Core)  

Reviewer: Core would not reject the service in this case, since this feature was not implemented at that time. Please change Core's response to `StartService ACK` in the diagram.

|||
Figure 4
![Compatibility NewProxy OldCore](./assets/Compatibility_NP_OC.png)
|||

### Backwards Compatibility(Old Proxy/New Core)  

|||
Figure 5
![Compatibility OldProxy NewCore](./assets/Compatibility_OP_NC.png)
|||

### TransportEventUpdate(Secondary Transport unavailable)  

|||
Figure 6
![TransportEventUpdate](./assets/TransportEventUpdate_Disconnected.png)
|||

# Setting up multiple transports

## Getting started

Add the following lines to `smartDeviceLink.ini`

> - To enable multiple transports in core

Reviewer: `Core` should probably be capitalized each time it is used.

```
[MultipleTransports]
...
MultipleTransportsEnabled = true
```

> - To set the allowed Secondary Transport types for a given Primary transport

Reviewer: ```suggestion
- To set the available Secondary Transport types for a given Primary transport
```

```
[MultipleTransports]
...
SecondaryTransportForBluetooth = WiFi
;SecondaryTransportForUSB =
;SecondaryTransportForWiFi =
```

> **List of secondary transport types**

Reviewer: This should likely be in the next section, since the `SecondaryTransportFor...` options use different values. Also, a note should be included with the values used in that configuration (`Wifi`, `Bluetooth`, `USB`).

| String | Description |
| ------ | ----------- |
|IAP_BLUETOOTH |	iAP over Bluetooth|
|IAP_USB |	iAP over USB, and Core cannot distinguish between Host Mode and Device Mode.|
|IAP_USB_HOST_MODE |	iAP over USB, and the phone is running as host|
|IAP_USB_DEVICE_MODE |	iAP over USB, and the phone is running as device|
|IAP_CARPLAY|	iAP over Carplay wireless|
|SPP_BLUETOOTH|	Bluetooth SPP. Either legacy SPP or SPP multiplexing.|
|AOA_USB|	Android Open Accessory|
|TCP_WIFI|	TCP connection over Wi-Fi|

## Audio and Video streaming

Modify the services map in `smartdeviceLink.ini` to restrict video and audio streaming services to specific transport types.

```
[ServicesMap]
...
AudioServiceTransports = TCP_WIFI
VideoServiceTransports = TCP_WIFI, AOA_USB
```
- Transports are listed in preffered order
- If a transport is not listed, then the service is not allowed to run on that transport

> - If the `AudioServiceTransports`/`VideoServiceTransports` line is omitted, service will be allowed to run on the primary transport

Reviewer: ```suggestion
- If the `AudioServiceTransports`/`VideoServiceTransports` line is omitted, service will be allowed to run on any transport
```

Author: Since there is no preferred transport type. Wouldn't the service attempt to start on the primary transport? If these lines are ommited the StartServiceACK would have 
```
audioServiceTransports : [1, 2]
videoServiceTransports : [1, 2]
```
